### ISSUE TITLE
`tracing_subscriber::fmt::time::UtcTime::rfc_3339()` has unpredictable output length

### ISSUE THREAD

**Issue by lilyball** [tokio-rs/tracing#1729]
### Version

```
asdf-tracing v0.1.0 (/Users/lily/Dev/Scratch/asdf-tracing)
├── tracing v0.1.29
│   ├── tracing-attributes v0.1.18 (proc-macro)
│   └── tracing-core v0.1.21
└── tracing-subscriber v0.3.1
    ├── tracing v0.1.29 (*)
    ├── tracing-core v0.1.21 (*)
    └── tracing-log v0.1.2
        └── tracing-core v0.1.21 (*)
```

### Platform

Darwin DeerBook 21.1.0 Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:23 PDT 2021; root:xnu-8019.41.5~1/RELEASE_X86_64 x86_64

### Crates

tracing-subscriber

### Description

When using `tracing-subscriber` with the `time` feature, the `fmt::time::UtcTime::rfc_3339()` timer provides no control over the number of subsecond digits it prints. It appears the implementation prints anywhere from zero to 9 digits such that there are no padding zeroes. What this means is that log lines printed using this timer will not have a consistent format.

To test this I ran the following:

```rust
use tracing::info;

fn main() {
    tracing_subscriber::fmt()
        .with_timer(tracing_subscriber::fmt::time::UtcTime::rfc_3339())
        .init();

    for _ in 1..100 {
        info!("this is a test");
    }
}
```

And the output looked like

```
2021-11-19T22:54:52.215771Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.21578Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215789Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215797Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215806Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215814Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215823Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215831Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.21584Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215848Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215857Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215865Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215874Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215883Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215891Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.2159Z  INFO asdf_tracing: this is a test
2021-11-19T22:54:52.215908Z  INFO asdf_tracing: this is a test
```

In my tests I've never dropped below 4 subsecond digits, but if I get extremely lucky it would even be possible to have a line that looks like 

```
2021-11-19T22:54:52Z  INFO asdf_tracing: this is a test
```

**Comment by lilyball** [comment]
As near as I can tell, `fmt::time::UtcTime::rfc_3339()` does pretty much the exact same thing as `fmt::time::time()` (which I believe is the default timer). `UtcTime` uses `time::OffsetDateTime::now_utc()` which seems to just call `std::time::SystemTime::now()` just like `fmt::time::time()` does. And tracing-subscriber's impl looks like it's doing the same RFC3339 format except with a fixed number of subsecond digits, whereas `time` is doing this format with a dynamic number of subsecond digits.

Which is to say, `fmt::time::UtcTime::rfc_3339()` appears to have no benefit over `fmt::time::time()`. You can't even use its `Default` impl to construct other `UtcTime` variants as there is no other `T: Formattable` that implements `Default` (and it's a sealed trait so you can't make your own). In theory, if the `time` crate adds other well-known formats, then this type can be used for those formats. But as it is today it doesn't do anything that tracing-subscriber can't already do natively.

**Comment by lilyball** [comment]
Presumably `LocalTime` has the same formatting issue. Simply deprecating `UtcTime` won't solve that.

The ideal solution is probably to file a ticket upstream with the `time` crate asking for a way to control the number of subsecond digits on the RFC3339 format, then update for that where `UtcTime::rfc_3339()` then defaults to 6 subsecond digits (and optionally provides an accessor to change that). This would be great, but is then blocked by the speed of development of the upstream `time` crate and how much they care.

Alternatively tracing-subscriber could take over the formatting of `UtcTime::rfc_3339()`. I'm not sure how to do this without a breaking change though, as right now it implements `FormatTime for UtcTime<T> where T: Formattable`, and it can't reasonably be specialized for `UtcTime<Rfc3339>`. I suppose it could in fact compare the `std::any::TypeId::of()` for the type and branch for `Rfc3339` specifically but that's, well, gross.

A third option is to just deprecate the `rfc_3339()` method in favor of a method that reconstructs the RFC3339 format with a fixed number of subsecond digits using `[FormatItem]`s. That's not very nice though. Or maybe it could just use its own type as the inner type and implement `FormatTime for UtcTime<NewType>` that just uses the existing built-in RFC3339 formatting.

**Comment by hawkw** [comment]
Zero-padding timestamps to the same width is probably worth an issue and/or pull request against the `time` crate. 

The `rfc_3339` constructor is provided primarily for completeness; the intention is to support any well-known formats that the `time` crate implements

> You can't even use its `Default` impl to construct other `UtcTime` variants as there is no other `T: Formattable` that implements `Default` (and it's a sealed trait so you can't make your own).

You can use [`UtcTime::new`](https://docs.rs/tracing-subscriber/0.3.1/tracing_subscriber/fmt/time/struct.UtcTime.html#method.new) to pass in a formatter constructed by parsing a user-provided format string, or produced at compile-time using the `time` crate's macros. That's the primary intention of the `time` crate integration; providing a way of parsing user-provided format specifiers is way out of scope for `tracing-subscriber`.

**Comment by lilyball** [comment]
Oops I meant to delete that paragraph.

**Comment by lilyball** [comment]
An upstream solution would be great. I guess the only real issue in tracing-subscriber right now is just that it provides a convenience accessor for a format that turns out to not have a constant length.

So perhaps in the short term it could just be documented with a note saying that it has a dynamic length due to subsecond formatting.

**Comment by Walther** [comment]
Hello hi! 👋 
What is the status of this - if an upstream fix is desired, does a tracking issue for that exist somewhere?
I tried to do a bit of [searching](https://github.com/time-rs/time/issues?q=is%3Aissue+length) - if this is the correct crate, I didn't find a respective issue. Should a new one be created?
Thanks!

**Comment by hawkw** [comment]
There should probably be an issue upstream if there isn't one already, so you're welcome to open one.

**Comment by Walther** [comment]
Upstream considers this as a wontfix:
>  jhpratt commented 28 minutes ago
> There is no guarantee of length beyond the fact that at least one digit will be present. This is consistent with the RFC3339 spec as written. It sounds like a custom format description should be used. This can guarantee a specific length.

https://github.com/time-rs/time/issues/440#issuecomment-1027846428

**Comment by lilyball** [comment]
I think the upstream comment here is right. `time`'s job isn't to produce a consistent length, it's simply to format. Ideally they'd have a variant of RFC3339 that takes a specific subsecond length, but for basic usage, minimizing unnecessary digits seems reasonable (e.g. if you're formatting into JSON, there's no reason to waste the bytes on trailing zeroes).

For tracing-subscriber's usage, using `time`'s RFC3339 format directly is going to be problematic because of this. Unless `time` introduces a variant with a fixed subsecond precision, `tracing-subscriber` should consider just providing its own RFC3339 format that does pick a specific (potentially customizable) precision.

**Comment by progre** [comment]
Glad to see that a PR for this issue is being created.
Hopefully they will be merged soon.
